### PR TITLE
LibVT+Terminal+LibLine: Implement line wrapping

### LINKED ISSUE #287: Terminal: Preservation and wrapping of terminal text when resizing terminal window
One of AK's commits from yesterday highlighted the issue that shrinking the terminal window is currently a destructive operation for any text it displays.

It would be great not to have to reprint ('re-cat') stuff after shrinking the terminal. This requires that the terminal and shell can auto-linewrap text when the window is being resized and the text cannot fit into the terminal column width.

### PR DESCRIPTION
**Pull request by alimpfard**
Fixes #287.
This is working.

History rewrap demo:

https://user-images.githubusercontent.com/14001776/122958386-0b1ef500-d398-11eb-90f0-b8463b5ac1ca.mp4

Things to figure out / fix:
- [x] Resizing to a small window with the buffer being full messes up the last line's placement
- [x] History is not re-wrapped on resize
- [x] LibLine can't handle having its origin changed, it should probably just change its origin to whatever the current line is when the resize happens
- [x] To deal with the above, the cursor should move with the lines:
- [x] Spill from the above, the cursor should move correctly when making the window smaller.

### CODE REVIEW

**Comment by awesomekling** [comment]
It might be easier to approach this in the same way as `GUI::TextEditor` works, with separation of physical lines and visual lines.

If you treat the terminal buffer as a text document, you can reflow the physical lines with different amounts of available size at will.

**Comment by alimpfard** [comment]
That'd be one big rewrite instead of two measly passes over the lines though
I'll take a look at it, but no promises

**Comment by alimpfard** [comment]
> If you treat the terminal buffer as a text document, you can reflow the physical lines with different amounts of available size at will.

Notes about this and why it's not how I implemented line wrapping, could save a future person trying to do it this way some brainpower:
- The terminal itself has to wrap, otherwise the cursor logic will be incorrect
- Since the terminal is wrapping, there's no real "physical line" available
- We _can_ go (logical terminal line) => (physical abstract line) => (logical widget line), but the terminal _has_ to have lines with correct lengths, otherwise the escape sequence logic will be out of touch with reality
- We could make the terminal keep physical lines and wrap it into logical lines, _then_ do the cursor/escape sequence logic, but then we'd have to rewrite quite a bit of the logic, and pay for the re-wrapping logic even when the terminal is not being resized

File: Userland/Libraries/LibVT/Line.cpp [hunk 1]
```diff
@@ -10,19 +10,86 @@ namespace VT {
...
+            if (!m_terminated_at.has_value()) {
+                auto cells_to_grab_from_next_line = min(new_length - old_length, next_line->length());
+                auto clear_next_line = false;
+                if (next_line->m_terminated_at.has_value()) {
+                    cells_to_grab_from_next_line = min(cells_to_grab_from_next_line, static_cast<size_t>(next_line->m_terminated_at.value()));
+                    if (cells_to_grab_from_next_line == *next_line->m_terminated_at) {
+                        m_terminated_at = old_length + *next_line->m_terminated_at;
+                        next_line->m_terminated_at.clear();
+                        clear_next_line = true;
+                    } else {
```
Comment (awesomekling) [inline Userland/Libraries/LibVT/Line.cpp:42]: I feel like this code is nested way more deeply than it needs to be.
Can we reorganize it a little bit and cut the nesting in ~half? :)

Comment (alimpfard) [reply Userland/Libraries/LibVT/Line.cpp:42]: I can try, but it's not immediately obvious to me how I would do that :thinking:

Comment (alimpfard) [reply Userland/Libraries/LibVT/Line.cpp]: I hope this nesting reduction is enough :P

Comment (awesomekling) [reply Userland/Libraries/LibVT/Line.cpp]: Thank you! Yes, I find it much more understandable now :)

**Pull request merged**